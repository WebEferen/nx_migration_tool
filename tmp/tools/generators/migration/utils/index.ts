export * from './directory';
export * from './transaction';
export * from './use-command';
