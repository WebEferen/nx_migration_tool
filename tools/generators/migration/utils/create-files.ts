import type { Tree } from '@nrwl/devkit';
import { readProjectConfiguration, generateFiles } from '@nrwl/devkit';
import path from 'path';

type PathType = { name: string; path: string };

export interface IOptions {
    targetName: string;
    targetPaths: PathType[];
    testPaths: PathType[];
}

export function createFiles(tree: Tree, options: IOptions) {
    const projectConfiguration = readProjectConfiguration(tree, options.targetName);
    const targetPaths = options.targetPaths || [];
    const testPaths = options.testPaths || [];

    generateFiles(tree, path.join(__dirname, 'templates'), projectConfiguration.root, {
        js: 'js',
        json: 'json',
        eslintrc: '.eslintrc',
        applicationName: options.targetName,
        // applicationPaths: [{ name: '@common', path: 'src/app/common' }],
        // testPaths: [{ name: '^@common$', path: './src/app/common' }],
        applicationPaths: targetPaths,
        testPaths: testPaths,
    });
}
