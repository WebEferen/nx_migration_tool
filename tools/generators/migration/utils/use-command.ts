import * as spawn from 'cross-spawn';

export async function useCommand(commandName: string, commandArgs: string[] = []) {
    const child = spawn(commandName, commandArgs);
    return new Promise<{ success: boolean }>((resolve) => {
        // child.stdout.on('data', (data) => process.stdout.write(data));
        child.stderr.on('data', (data) => process.stderr.write(data));

        child.on('close', (code) => resolve({ success: code === 0 }));
        child.on('error', () => resolve({ success: false }));
    });
}
