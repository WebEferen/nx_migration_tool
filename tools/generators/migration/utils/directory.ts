import { useCommand } from './use-command';

export async function manageDirectories(tempDir: string, targetDir: string) {
    await useCommand('rm', ['-rf', tempDir]);
    await useCommand('mkdir', [tempDir]);

    await useCommand('rm', ['-rf', `apps/${targetDir}`]);

    await useCommand('mkdir', [`apps/${targetDir}`]);
    await useCommand('mkdir', [`apps/${targetDir}/_to_manual_check_`]);
}
