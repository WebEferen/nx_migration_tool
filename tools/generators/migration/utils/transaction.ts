import * as uuid from 'uuid';

import { useCommand } from './use-command';

export async function rollbackTransaction(branchName: string, fallbackBranchName: string) {
    await useCommand('git', ['reset', '--hard']);
    await useCommand('git', ['checkout', fallbackBranchName]);
    await useCommand('git', ['branch', '-D', branchName]);
    await useCommand('rm', ['-rf', branchName]);

    throw new Error('Error during transaction... Aborting and rolling back (git-reset)');
}

export async function useTransaction(callback: (branchName: string) => Promise<void>) {
    const branchName: string = uuid.v4().replace(/-/g, '').slice(0, 16);

    const branchChanged = await useCommand('git', ['checkout', '-b', branchName]);
    if (branchChanged.success) await callback(branchName).catch((_) => process.stdout.write('Error during transaction!'));
    if (branchChanged.success) await useCommand('git', ['branch', '-D', branchName]);
}
