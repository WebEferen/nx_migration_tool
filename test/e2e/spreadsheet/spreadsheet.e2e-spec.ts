import { agent, generateTokenForXsuaa } from '../utils';
import invalidPayloadInputString from './__fixtures__/payload-empty-workbook-input-string.fixture.json';
import invalidPayloadEmptyWorkbook from './__fixtures__/payload-empty-workbook.fixture.json';
import fallbackPayloadWithCellFormulas from './__fixtures__/payload-fallback-with-cell-formulas.fixture.json';
import fallbackPayloadWithDislocatedTable from './__fixtures__/payload-fallback-with-dislocated-table-workbook.fixture.json';
import fallbackPayloadNoFieldInDataSource from './__fixtures__/payload-fallback-workbook-no-field-in-data-source.fixture.json';
import fallbackPayloadNoFieldInWorkbook from './__fixtures__/payload-fallback-workbook-no-field-in-workbook.fixture.json';
import fallbackPayload from './__fixtures__/payload-fallback-workbook.fixture.json';
import fullPayloadNoFieldInWorkbookResults from './__fixtures__/payload-full-workbook-no-field-in-workbook-results.fixture.json';
import fullPayload from './__fixtures__/payload-full-workbook.fixture.json';
import invalidPayload from './__fixtures__/payload-invalid-workbook.fixture.json';

describe('Spreadsheet calculation', () => {
    describe('/api/spreadsheet/calculate (POST)', () => {
        const tokenPayload = { cid: '12345', azp: '12345', zid: 'tenant-id' };
        const validTokenWithoutScope = generateTokenForXsuaa(tokenPayload);
        const validToken = generateTokenForXsuaa({ scope: ['papm.run_calculation'], ...tokenPayload });

        it('should correctly calculate formulas from _INPUT table based on input and lookup data', async () => {
            // given

            // when
            const response = agent.post(`/api/spreadsheet/calculate`).set('Authorization', `Bearer ${validToken}`).send(fallbackPayload);

            // then
            const results = await response
                .expect('Content-Type', /json/)
                .expect(200)
                .then<any>(({ body }) => body);

            expect(results).toMatchSnapshot();
        });

        it('should correctly calculate formulas from dislocated _INPUT table based on input and lookup data', async () => {
            // given

            // when
            const response = agent
                .post(`/api/spreadsheet/calculate`)
                .set('Authorization', `Bearer ${validToken}`)
                .send(fallbackPayloadWithDislocatedTable);

            // then
            const results = await response
                .expect('Content-Type', /json/)
                .expect(200)
                .then<any>(({ body }) => body);

            expect(results).toMatchSnapshot();
        });

        it('should correctly calculate formulas from _RESULT table based on input and lookup data', async () => {
            // given

            // when
            const response = agent.post(`/api/spreadsheet/calculate`).set('Authorization', `Bearer ${validToken}`).send(fullPayload);

            // then
            const results = await response
                .expect('Content-Type', /json/)
                .expect(200)
                .then<any>(({ body }) => body);

            expect(results).toMatchSnapshot();
        });

        it('should fail with "400 Bad Request" when sending empty payload', async () => {
            // given
            const emptyPayload = {};

            // when
            const response = agent.post(`/api/spreadsheet/calculate`).set('Authorization', `Bearer ${validToken}`).send(emptyPayload);

            // then
            await response.expect(400);
        });

        it('should fail with "400 Bad Request" when sending invalid input data', async () => {
            // given
            const payloadWithInvalidInputData = { data: '', workbook: '' };

            // when
            const response = agent.post(`/api/spreadsheet/calculate`).set('Authorization', `Bearer ${validToken}`).send(payloadWithInvalidInputData);

            // then
            await response.expect(400);
        });

        it('should fail with "422 Unprocessable Entity" when sending invalid workbook', async () => {
            // given

            // when
            const response = agent.post(`/api/spreadsheet/calculate`).set('Authorization', `Bearer ${validToken}`).send(invalidPayload);

            // then
            await response.expect(422);
        });

        it('should correctly return data based on _INPUT table with many formulas and formatters', async () => {
            // given

            // when
            const response = agent
                .post(`/api/spreadsheet/calculate`)
                .set('Authorization', `Bearer ${validToken}`)
                .send(fallbackPayloadWithCellFormulas);

            // then
            const results = await response
                .expect('Content-Type', /json/)
                .expect(200)
                .then<any>(({ body }) => body);

            expect(results).toMatchSnapshot();
        });

        it('should fail when token is not valid', async () => {
            // given
            const invalidAuthToken = Buffer.from('FOOBAR').toString('base64');

            // when
            const request = agent
                .post('/api/spreadsheet/calculate')
                .set('Authorization', `Bearer ${invalidAuthToken}`)
                .send(fallbackPayloadWithCellFormulas);

            //then
            await request.expect(401);
        });

        it('should return "Forbidden resource" when unauthorized', async () => {
            // given

            // when
            const request = agent
                .post('/api/spreadsheet/calculate')
                .set('Authorization', `Bearer ${validTokenWithoutScope}`)
                .send(fallbackPayloadWithCellFormulas);

            //then
            await request.expect(403);
        });

        it('should return status 422 for incorrect _INPUT table columns in data source', async () => {
            // given

            // when
            const response = agent
                .post(`/api/spreadsheet/calculate`)
                .set('Authorization', `Bearer ${validToken}`)
                .send(fallbackPayloadNoFieldInDataSource);

            // then
            const results = await response
                .expect('Content-Type', /json/)
                .expect(422)
                .then<any>(({ body }) => body);

            expect(results.message).toMatchSnapshot();
        });

        it('should return status 422 for incorrect _INPUT table columns in workbook', async () => {
            // given

            // when
            const response = agent
                .post(`/api/spreadsheet/calculate`)
                .set('Authorization', `Bearer ${validToken}`)
                .send(fallbackPayloadNoFieldInWorkbook);

            // then
            const results = await response
                .expect('Content-Type', /json/)
                .expect(422)
                .then<any>(({ body }) => body);

            expect(results.message).toMatchSnapshot();
        });

        it('should return status 422 for incorrect _RESULT table columns in workbook', async () => {
            // given

            // when
            const response = agent
                .post(`/api/spreadsheet/calculate`)
                .set('Authorization', `Bearer ${validToken}`)
                .send(fullPayloadNoFieldInWorkbookResults);

            // then
            const results = await response
                .expect('Content-Type', /json/)
                .expect(422)
                .then<any>(({ body }) => body);

            expect(results.message).toMatchSnapshot();
        });

        it('should return status 422 for incorrect data source as string instead of array', async () => {
            // given

            // when
            const response = agent.post(`/api/spreadsheet/calculate`).set('Authorization', `Bearer ${validToken}`).send(invalidPayloadInputString);

            // then
            const results = await response
                .expect('Content-Type', /json/)
                .expect(422)
                .then<any>(({ body }) => body);

            expect(results.message).toMatchSnapshot();
        });

        it('should return status 422 when table not found in wokbook', async () => {
            // given

            // when
            const response = agent.post(`/api/spreadsheet/calculate`).set('Authorization', `Bearer ${validToken}`).send(invalidPayloadEmptyWorkbook);

            // then
            const results = await response
                .expect('Content-Type', /json/)
                .expect(422)
                .then<any>(({ body }) => body);

            expect(results.message).toMatchSnapshot();
        });
    });
});
