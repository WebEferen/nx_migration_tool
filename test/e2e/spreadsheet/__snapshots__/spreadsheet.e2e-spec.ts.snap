// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spreadsheet calculation /api/spreadsheet/calculate (POST) should correctly calculate formulas from _INPUT table based on input and lookup data 1`] = `
Object {
  "data": Array [
    Object {
      "AMOUNT": " 987.00 € ",
      "CURRENCY": "EUR",
      "FISCYEARPER": "3/30/2020",
      "GLACC": "A1",
    },
    Object {
      "AMOUNT": " 876.00 € ",
      "CURRENCY": "EUR",
      "FISCYEARPER": "3/30/2020",
      "GLACC": "A2",
    },
    Object {
      "AMOUNT": " 765.00 € ",
      "CURRENCY": "EUR",
      "FISCYEARPER": "3/30/2020",
      "GLACC": "A3",
    },
    Object {
      "AMOUNT": " 654.00 € ",
      "CURRENCY": "EUR",
      "FISCYEARPER": "3/30/2020",
      "GLACC": "A4",
    },
    Object {
      "AMOUNT": " 543.00 € ",
      "CURRENCY": "EUR",
      "FISCYEARPER": "3/30/2020",
      "GLACC": "A5",
    },
    Object {
      "AMOUNT": " 2.00 € ",
      "CURRENCY": "EUR",
      "FISCYEARPER": "3/30/2020",
      "GLACC": "A6",
    },
    Object {
      "AMOUNT": " 2.00 € ",
      "CURRENCY": "EUR",
      "FISCYEARPER": "3/30/2020",
      "GLACC": "A7",
    },
  ],
}
`;

exports[`Spreadsheet calculation /api/spreadsheet/calculate (POST) should correctly calculate formulas from _RESULT table based on input and lookup data 1`] = `
Object {
  "data": Array [
    Object {
      "AMOUNT": "987",
      "CURRENCY": "EUR",
      "FISCYEARPER": "4/30/2020",
      "GLACC": "A1",
      "PRODUCT": "P2",
      "QUANTITY": "3.0",
      "UNIT": "PCS",
    },
    Object {
      "AMOUNT": "500.571428571429",
      "CURRENCY": "EUR",
      "FISCYEARPER": "3/31/2020",
      "GLACC": "A2",
      "PRODUCT": "P1",
      "QUANTITY": "4.0",
      "UNIT": "PCS",
    },
    Object {
      "AMOUNT": "292",
      "CURRENCY": "EUR",
      "FISCYEARPER": "3/31/2020",
      "GLACC": "A2",
      "PRODUCT": "P2",
      "QUANTITY": "2.0",
      "UNIT": "PCS",
    },
    Object {
      "AMOUNT": "255",
      "CURRENCY": "EUR",
      "FISCYEARPER": "3/31/2020",
      "GLACC": "A3",
      "PRODUCT": "P1",
      "QUANTITY": "1.0",
      "UNIT": "PCS",
    },
    Object {
      "AMOUNT": "510",
      "CURRENCY": "EUR",
      "FISCYEARPER": "3/31/2020",
      "GLACC": "A3",
      "PRODUCT": "P2",
      "QUANTITY": "2.0",
      "UNIT": "PCS",
    },
    Object {
      "AMOUNT": "218",
      "CURRENCY": "EUR",
      "FISCYEARPER": "3/31/2020",
      "GLACC": "A4",
      "PRODUCT": "P1",
      "QUANTITY": "1.0",
      "UNIT": "PCS",
    },
    Object {
      "AMOUNT": "436",
      "CURRENCY": "EUR",
      "FISCYEARPER": "3/31/2020",
      "GLACC": "A4",
      "PRODUCT": "P2",
      "QUANTITY": "2.0",
      "UNIT": "PCS",
    },
    Object {
      "AMOUNT": "181",
      "CURRENCY": "EUR",
      "FISCYEARPER": "3/31/2020",
      "GLACC": "A5",
      "PRODUCT": "P1",
      "QUANTITY": "1.0",
      "UNIT": "PCS",
    },
    Object {
      "AMOUNT": "362",
      "CURRENCY": "EUR",
      "FISCYEARPER": "3/31/2020",
      "GLACC": "A5",
      "PRODUCT": "P2",
      "QUANTITY": "2.0",
      "UNIT": "PCS",
    },
  ],
}
`;

exports[`Spreadsheet calculation /api/spreadsheet/calculate (POST) should correctly calculate formulas from dislocated _INPUT table based on input and lookup data 1`] = `
Object {
  "data": Array [
    Object {
      "AMOUNT": " 987.00 € ",
      "CURRENCY": "EUR",
      "FISCYEARPER": "3/30/2020",
      "GLACC": "A1",
    },
    Object {
      "AMOUNT": " 876.00 € ",
      "CURRENCY": "EUR",
      "FISCYEARPER": "3/30/2020",
      "GLACC": "A2",
    },
    Object {
      "AMOUNT": " 765.00 € ",
      "CURRENCY": "EUR",
      "FISCYEARPER": "3/30/2020",
      "GLACC": "A3",
    },
    Object {
      "AMOUNT": " 654.00 € ",
      "CURRENCY": "EUR",
      "FISCYEARPER": "3/30/2020",
      "GLACC": "A4",
    },
    Object {
      "AMOUNT": " 543.00 € ",
      "CURRENCY": "EUR",
      "FISCYEARPER": "3/30/2020",
      "GLACC": "A5",
    },
    Object {
      "AMOUNT": " 2.00 € ",
      "CURRENCY": "EUR",
      "FISCYEARPER": "3/30/2020",
      "GLACC": "A6",
    },
    Object {
      "AMOUNT": " 2.00 € ",
      "CURRENCY": "EUR",
      "FISCYEARPER": "3/30/2020",
      "GLACC": "A7",
    },
  ],
}
`;

exports[`Spreadsheet calculation /api/spreadsheet/calculate (POST) should correctly return data based on _INPUT table with many formulas and formatters 1`] = `
Object {
  "data": Array [
    Object {
      "CC": "CC1",
      "FA": "FA1",
      "GL": "GL1",
      "N1": "100",
      "N2": "10.00",
      "N3": "1,001.00 USD",
      "PC": "PC1",
    },
    Object {
      "CC": "CC2",
      "FA": "FA1",
      "GL": "GL1",
      "N1": "200",
      "N2": "10.00",
      "N3": "2,001.00 USD",
      "PC": "PC1",
    },
    Object {
      "CC": "CC1",
      "FA": "FA2",
      "GL": "GL1",
      "N1": "300",
      "N2": "20.00",
      "N3": "6,001.00 USD",
      "PC": "PC2",
    },
    Object {
      "CC": "CC2",
      "FA": "FA2",
      "GL": "GL1",
      "N1": "400",
      "N2": "20.00",
      "N3": "8,001.00 USD",
      "PC": "PC2",
    },
    Object {
      "CC": "CC1",
      "FA": "FA1",
      "GL": "GL2",
      "N1": "500",
      "N2": "10.00",
      "N3": "5,001.00 USD",
      "PC": "PC1",
    },
    Object {
      "CC": "CC2",
      "FA": "FA1",
      "GL": "GL2",
      "N1": "600",
      "N2": "10.00",
      "N3": "6,001.00 USD",
      "PC": "PC1",
    },
    Object {
      "CC": "CC1",
      "FA": "FA2",
      "GL": "GL2",
      "N1": "700",
      "N2": "20.00",
      "N3": "14,001.00 USD",
      "PC": "PC2",
    },
    Object {
      "CC": "CC2",
      "FA": "FA2",
      "GL": "GL2",
      "N1": "800",
      "N2": "20.00",
      "N3": "16,001.00 USD",
      "PC": "PC2",
    },
  ],
}
`;

exports[`Spreadsheet calculation /api/spreadsheet/calculate (POST) should return status 422 for incorrect _INPUT table columns in data source 1`] = `"Fields in table '_INPUT' don't match. Fields ['Formel'] don't exist in data source."`;

exports[`Spreadsheet calculation /api/spreadsheet/calculate (POST) should return status 422 for incorrect _INPUT table columns in workbook 1`] = `"Fields in table '_INPUT' don't match. Fields ['FISCYEARPER'] don't exist in workbook."`;

exports[`Spreadsheet calculation /api/spreadsheet/calculate (POST) should return status 422 for incorrect _RESULT table columns in workbook 1`] = `"Fields in table '_RESULT' don't match. Fields ['FISCYEARPER'] don't exist in workbook."`;

exports[`Spreadsheet calculation /api/spreadsheet/calculate (POST) should return status 422 for incorrect data source as string instead of array 1`] = `"Input data for table '_INPUT' is not an array, string was passed instead."`;

exports[`Spreadsheet calculation /api/spreadsheet/calculate (POST) should return status 422 when _INPUT or _RESULT tables not found in workbook 1`] = `"Couldn't find result tables '_RESULT' and '_INPUT' in provided workbook"`;

exports[`Spreadsheet calculation /api/spreadsheet/calculate (POST) should return status 422 when table not found in workbook 1`] = `"Table '_INPUT' doesn't exist in workbook"`;
